
% Notes on modeling heart rate

\documentclass[12pt]{article}
\usepackage{graphicx,color}
\usepackage{amsmath, amsfonts}
\usepackage{placeins}
\usepackage{verbatim}
\usepackage{showlabels}

\newcommand{\field}[1]{\mathbb{#1}}
\newcommand{\INTEGER}{\field{Z}}
\newcommand{\REAL}{\field{R}}
\newcommand{\COMPLEX}{\field{C}}
\newcommand{\id}{\mathbb{I}}
\newcommand{\variance}[1]{\field{V}\left[ #1 \right]}
\newcommand{\normal}[2]{{\cal N}\left(#1,#2 \right)}
\newcommand{\argmax}{\operatorname*{argmax}}
\newcommand{\BestModel}{\emph{two\_ar5\_masked}}

\title{HMMs for Heart Rate}
\author{Andrew M.\ Fraser}

\begin{document}
\maketitle

\section{Introduction}
\label{sec:introduction}

Here I document ideas for redoing the last chapter of my book
\emph{Hidden Markov Models and Dynamical Systems} in which I address
the CINC-2000 challenge.  I want to map the products of the ECG
analysis described in my document \emph{HMMs for ECGs} to the two
classification challenges.  The inputs are time series of heart rate
estimates and likelihoods sampled at 2 Hz and the outputs are two
stages of classification.  Pass-1: For each record determine whether
or not the patient experienced apnea; and Pass-2: for each minute of
each record classified as apnea in pass-1, determine whether or not
the patient was experiencing apnea.

\section{A Model for Record a03}
\label{sec:a03}

\textbf{To Do:}  Describe structure of the model and its performance.

\subsection{C records}
\label{sec:c_records}

Because the C records have only a few or zero minutes marked as apnea,
trying to create and train models with states for the apnea class
fails.  I build models of C records that have a single state and that
state is in the normal class.

\section{Pass-1 Challenges}
\label{sec:pass-1}

I built a model called \BestModel with the following
features:
\begin{itemize}
\item Models detected peaks and forces them to be on chains
\item There are \emph{two} chains: One for normal; And one for apnea
\item The lengths of the \emph{intervals} between peaks are modeled
\item The heart rate signal is fit to AR5 models
\end{itemize}
Here I will refer to that model as simply the \emph{HMM}.

The error rate for pass-2 with the HMM on the \emph{a} records is
lower than 0.5 except for \emph{a05} \emph{a09} \emph{a18} and
\emph{a20}.  I'd hoped that I could rely on high likelihood to
identify problematic \emph{x} records and mark them as all apnea, but
that doesn't work.

\begin{figure}
  \centering
    \resizebox{0.8\textwidth}{!}{\includegraphics{statistics.pdf}}
    \caption{Pass-1 statistic.  Power spectral density (PSD) estimates
      appear in the upper plots.  The solid curves are averages over
      the \emph{a} and \emph{c} records (in red and blue
      respectively).  In the lower plots, the y values are, $G(PSD)$,
      which is the sum of all channels above 0.3 cpm.  For each
      record, I divide the raw PSD by $G(PSD)$ and then sum the
      channels from 1.0 to 3.6 cpm to obtain the pass-1 statistic,
      $F(PSD)$, which is plotted on the x axis.  In pass 2, I treat
      records that are left of the vertical dotted line as normal and
      don't classify any of the minutes as apnea.  The only training
      record that pass-1 misclassifies is \emph{c10}, and pass-2
      performs well on that record.}
  \label{fig:statistics}
\end{figure}

Figure~\ref{fig:statistics}\footnote{The PSDs are normalized to have
  the same power for frequencies above 0.3 cpm)} illustrates pass-1.
  The following \emph{x} records fall close to the boundary:
\begin{description}
\item[x06] Pass-1 says A which seems \textbf{wrong}.  But HMM says this
  is mostly normal, which seems correct.
\item[x09] Pass-1 says A which seems correct, but HMM says this is
  mostly normal.
\item[x10] Pass-1 says A which seems correct.  HMM misses some of the
  apnea
\item[x13] Pass 1 says A which seems correct.  HMM misses some of the
  apnea.
\item[x15] Pass-1 says A.  Maybe it's B.  HMM results seem plausible.
\item[x17] Pass-1 says A which seems \textbf{wrong}, but hmm says this
  is mostly normal.
\item[x23] Pass-1 says A which seems correct.  HMM has trouble
  detecting apnea.
\item[x24] Pass-1 says N which seems correct.  HMM says almost all
  normal.
\item[x33 \& x34] Pass-1 says N.  I don't know if that's correct.
  Visually I guess there might be a few episodes of apnea.  HMM
  agrees.  I think \emph{x33} and \emph{x34} are different electrodes
  measuring the same record.
\end{description}

Here is my guess at classification of the records:
\begin{description}
\item[C]
  x03 (classified as A),
  x04,
  x06,
  x11,
  x17,
  x18 (classified as A),
  x22,
  x24,
  x29,
  x35
\item[B]
  x13 (A),
  x15 (A),
  x26 (N),
  x33 (N),
  x34 (N)
\item[A]
  x01, x02, x05, x07, x08, x09, x10, x12, x14, x16, x19, x20,
  x21, x23, x25, x27, x28, x30, x31, x32
\end{description}

x29 suffers alternans not apnea.

Setting the threshold for statistic 1 to 0.36 classifies all of the
\emph{a} and \emph{c} records records correctly except c10, \emph{and}
the way it classifies the \emph{x} records is consistent with my
visual inspection.

These are the records for which are close to the boundary for pass-1:
\begin{description}
\item[c09] Like c02.  Hour to hour heart rate changes.
\item[c07] Low hour to hour heart rate variation.
\item[x24]  Need to trim first 15 minutes.  Looks normal to me.
\item[x33 x34] These are different measurements of the same thing.
  I think it's a B record.
\item[b04] HMM correctly says mostly normal.
\item[c08] HMM says some apnea, but expert says none.
\item[c02] Hour to hour heart rate changes.
\item[x06] First 5.2 minutes need to be trimmed.  I guess it's a B
  record.
\item[x17] Last 5 minutes need to be trimmed.  Apnea 126-150.
\item[x18] Need to trim first 14.9 minutes.  206-230 looks like apnea
  oscillations, but respiration look sort of OK.
\item[x15] Apnea 120 - 140, 180 - 210, 350-390, 490-515.  More hour to
  hour variation than c07.
\item[a11] PMD for HMM is .64
\item[a06] Amplitude of apnea related heart rate oscillations is
  small.  The hour to hour changes may be helpful.  Classifying the
  minutes looks hard.
\item[x10] Apnea 220 - 250, 360-390
\item[x13] I guess it's B.
\item[x07] I see apnea oscillations.
\item[a09] Respiration signal is very weak.  Heart rate changes hour
  to hour but not as much as a10.  The amplitude of the oscillations
  in the heart rate during apnea is small.  It seems difficult to me.
\item[c10] Heart rate has apnea like oscillations but returns to same
  baseline.  During those oscillations the respiration signal also
  oscillates, but perhaps not as deeply as a10.
\item[a10] Heart rate changes hour to hour
\item[x09] HMM misses clear apnea episodes
\item[x23] Apnea 120-185 but respiration looks sort of OK
\end{description}

\textbf{Conclude:} Statistic 1 alone is adequate for pass-1.

\section{Work on Pass-2}
\label{sec:pass-2}

I want pass-2 code that will classify each minute of a heart rate
signal as either \emph{apnea} or \emph{normal}.

\subsection{Features}
\label{sec:reatures}

Here I look at features that I might use for classifying minutes of
heart rate signals.

\begin{figure}
  \centering
    \resizebox{0.8\textwidth}{!}{\includegraphics{analyze_peaks.pdf}}
    \caption{The distribution of local peaks in heart rate signals.}
  \label{fig:analyze_peaks}
\end{figure}

\begin{figure}
  \centering
    \resizebox{0.8\textwidth}{!}{\includegraphics{interval_pdfs.pdf}}
    \caption{Result of density ratio estimation.}
  \label{fig:interval_pdfs}
\end{figure}

\subsection{Setting Parameters}
\label{sec:setting_parameters}

Set the following parameters:
\begin{description}
\item[Minimum prominence $6.0$:] For detecting peaks.  See
  Fig.~\ref{fig:prominence_study}.  Available as args.min\_prominence
  from utilities.py, and specified as \emph{MIN\_PROMINENCE} in Makefile.
\item[Power $2.0$:] Exponent for weighting the \emph{interval}
  component of observations.  See Fig.~\ref{fig:power_threshold}.
  Available as args.power\_and\_threshold from utilities.py
\item[Threshold $1.0$:] For detecting apnea.  Larger values decreases
  false alarms and increases missed detections.  See
  Fig.~\ref{fig:power_threshold}.  Available as
  args.power\_and\_threshold from utilities.py
\item[AR Order $5$:] Number of past filtered heart rate samples used
  by observation model to forecast next sample.  See
  Fig.~\ref{fig:errors_vs_ar_order}.  Available as args.AR\_order from
  utilities.py, and specified as \emph{BEST\_MODEL} in Makefile.
\end{description}

\begin{figure}
  \centering
    \resizebox{0.8\textwidth}{!}{\includegraphics{prominence_study.pdf}}
    \caption{Error rate on the \emph{a} records for models with
      varying prominence threshold with form
      \emph{two\_ar5\_intervals*} .}
  \label{fig:prominence_study}
\end{figure}

\begin{figure}
  \centering
    \resizebox{0.8\textwidth}{!}{\includegraphics{power_threshold.pdf}}
    \caption{Error rate of model vs the weighting exponent and the
      detection threshold with model \BestModel on the
      \emph{a} records.}
  \label{fig:power_threshold}
\end{figure}

\begin{figure}
  \centering
    \resizebox{0.8\textwidth}{!}{\includegraphics{errors_vs_ar_order.pdf}}
    \caption{A plot of error rate vs.\ AR order for the training
      records in group \emph{A}.  The models have two chains for
      peaks; one chain for normal peaks and one chain for apnea
      peaks.  The performance on records in group \emph{C} is worse.
      Perhaps overall performance would be better if the threshold for
      detection were higher.}
  \label{fig:errors_vs_ar_order}
\end{figure}

\begin{table*}
  \centering
  \input{score.tex}
  \caption[Performance]{Performance of pass-1 combined with pass-2 on
    training data with model
    \BestModel\ for pass-2.}
  \label{tab:score_initial}
\end{table*}

\subsection{Normalized Two Chain Models}
\label{sec:norm_two}


\begin{figure}
  \centering
    \resizebox{0.8\textwidth}{!}{\includegraphics{norm_analyze_peaks.pdf}}
    \caption{The distribution of local peaks in \textbf{normalized}
      heart rate signals.}
  \label{fig:norm_analyze_peaks}
\end{figure}

\begin{figure}
  \centering
    \resizebox{0.8\textwidth}{!}{\includegraphics{norm_interval_pdfs.pdf}}
    \caption{Result of density ratio estimation for the time between peaks.}
  \label{fig:norm_interval_pdfs}
\end{figure}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\newcommand{\StudyFigs}[2]{
\begin{figure}
  \centering
    \resizebox{0.8\textwidth}{!}{\includegraphics{#1_prominence_study.pdf}}
    \caption{Like Fig.~\ref{fig:prominence_study} but for models #2.}
  \label{fig:#1_prominence_study}
\end{figure}

\begin{figure}
  \centering
  \resizebox{0.8\textwidth}{!}{\includegraphics{#1_power_threshold.pdf}}
  \caption{Like Fig.~\ref{fig:power_threshold} but for a model #2.}
  \label{fig:#1_power_threshold}
\end{figure}

\begin{figure}
  \centering
  \resizebox{0.8\textwidth}{!}{\includegraphics{#1_errors_vs_ar_order.pdf}}
    \caption{Like Fig.~\ref{fig:errors_vs_ar_order} but for models
      #2.}
  \label{fig:#1_errors_vs_ar_order}
\end{figure}

\begin{table*}
  \centering
  \input{#1_score.tex}
  \caption[Performance]{Like Table~\ref{tab:score_initial} but for a
    model #2.}
  \label{tab:#1_score}
\end{table*}}

\StudyFigs{norm}{with \textbf{normalized} heart rate signals}

\subsection{Respiration from Heart Rate}
\label{sec:respiration}

A normal heart rate signal is modulated by
respiration,\marginpar{illustrate with a figure} and I've written code
to extract a time series of the amplitude of that modulation.
Combining that respiration time series with a low pass filtered heart
rate signal yields a vector time series.
Table~\ref{tab:lphr_respiration_score} reports the classification
performance of a two state HMM with AR order 6 output models for such
vector time series.
\begin{table*}
  \centering
  \input{lphr_respiration_score.tex}
  \caption[Performance]{Like Table~\ref{tab:score_initial} but for a
    two state vector autoregressive model of the filtered heart rate
    and a \textbf{respiration} signal derived from the heart rate.}
  \label{tab:lphr_respiration_score}
\end{table*}

\subsection{Combining Intervals, Respiration and Heart Rate}
\label{sec:combination}

Next, I combine the vector heart rate and respiration signal of
Section~\ref{sec:respiration} with the two chain state structure and
interval modeling of Section~\ref{sec:norm_two}.  The parameters to
choose include:
\begin{description}
\item[VARG parameters:] for the (Vector AutoRegressive Gaussian) model
  of the 2-d vectors with components respiration and low-pass filtered
  heart rate.  The parameters include:
  \begin{description}
  \item[sample rate] 
  \item[normalization] 
  \item[low pass width] 7.5 per minute
  \item[respiration pass center] 15 per minute
  \item[respiration pass width] 3 per minute
  \item[envelope smooth] 1.5 per minute
  \item[AR order] 
  \item[$\psi$ and $\nu$] Priors for each state
  \end{description}
\item[Peak parameters] For finding peaks.  Note: The signal used is
  from the low pass filter for the VARG signal.
  \begin{description}
  \item[minimum prominence] 
  \item[distance] 0.417 minutes
  \item[window length] 1.42 minutes
  \end{description}
\item[Detector threshold] 
\item[Weighting power] 
\end{description}

\subsection{To Do}

\begin{enumerate}
\item Explore how performance depends on the parameters of the model
  in Sec.~\ref{sec:combination}.
\end{enumerate}

\subsubsection{Problematic Records}
\label{sec:problematic_records}

In Table~\ref{tab:norm_score} These are the records with more than 165
errors:
\begin{description}
\item[a05 229] Complex oscillations (period 2 ish, maybe multi-chain would
  fix) Should be possible to detect apnea that starts at 159.
  \textbf{fixable}
\item[a06 175] 360-370 looks \textbf{hopeless}.  I just don't see apnea vs
  normal in the heart rate signal.
\item[a15 178] Apnea starts at 33 and I don't see it in the heart rate
  signal.  It looks \textbf{hopeless} to me.
\item[a20 210] Apnea starts at 50 and seems visible in HR signal,  The
  oscillations are slow (.9 cpm).  At 355 oscillations are complex
  (multi period).  \textbf{Fixable}
\item[b02 267] 
\end{description}

\section{Synchronization}
\label{sec:sync}

I use several time series in this project.  After the modifications I
made in preparing this section, I believe that $t_0$ means the same
thing for each of them.  I was particularly worried about the alignment
of the expert annotations and the annotations estimated using a model.
Here are the functions that calculate the estimated annotations:
\begin{description}
\item[pass2.analyze] Calls utilities.Score2(model\_path, name)
\item[utilities.Score2.\_\_init\_\_] Reads a model (develop.HMM) and
  uses the develop.HMM.read\_y\_no\_class method (which calls
  args.read\_raw\_y(args, record\_name)) to read the data.
\item[model\_init.two\_intervals] Assigns
  utilities.read\_slow\_peak\_interval to args.read\_raw\_y
\item[utilities.read\_slow\_peak\_interval] Calls functions to add
  intervals and peaks to the result of calling
  utilities.read\_slow(args, name)
\item[utilities.add\_intervals] 
\item[utilities.add\_peaks]
\item[utilities.read\_slow] Calls
  utilities.read\_slow\_fast\_respiration(args, name) and then returns
\begin{verbatim}
{'slow': input_['slow'][trim_samples:-trim_samples]}
\end{verbatim}
\item[utilities.read\_slow\_fast\_respiration] Assigns int(
  (args.heart\_rate\_sample\_frequency * args.trim\_start)) to
  result['trim\_samples']
\item[utilities.common\_arguments] The default for trim\_start is 0
\end{description}

Here are the lengths of some time series for each record in the
training data:

\verbatiminput{lengths.txt}

After looking at a preliminary version of these lengths, I added
padding to the beginning of the \emph{slow} time series to match the
AR order of the \emph{slow} model.  Now, for an AR-5 model, they are
12.5 seconds longer than the \emph{ecg} time series.  I also estimate
an extra minute for any record that has ecg data that doesn't end on
an even minute.

\section{Unlucky Thoughts}
\label{sec:unlucky}

Here are some ideas that didn't work out\footnote{He's not stupid.
  He's just unlucky when he thinks.}:
\begin{description}
\item[Multi-chain] Using different chains for different peak prominences.
\item[Cumulative distribution of respiration] I looked at the CDF of
  the respiration signal.  I hoped to use it to resolve records in
  pass-1 that the low frequency PSD left on the border.
\item[Use ECG Likelihood] My idea was to use the likelihood of the ECG
  model to modify the variances of the heart rate models.  I didn't
  see any structure to exploit when I looked at plots of the log
  likelihood, $\log\left(P(y[t]|y[:t]) \right)$, of the heart rate
  model against the log likelihood of the ECG model.
\item[Multiple Chains] In model\_init.py I made chains with lengths
  numpy.arange(20, 40, 2) for both normal and apnea.  Initially the
  models had AR order 1.  After writing code to change the AR order, I
  found the log likelihood per time step varied radically between
  records.  Here is the last line from training an AR-4 model for 20
  iterations:
\begin{verbatim}
19 L0 -5.875e+03 L1  6.323e+03 L2  1.646e+03 L3  5.682e+03 L4  4.964e+03
L5 -1.093e+04 L6  3.797e+03 L7  1.924e+03 L8 -4.514e+03 L9  4.667e+03
L10 -1.755e+03 L11  5.678e+03 L prior -9.081e+06 U/n -4.1274e+01
\end{verbatim}
  I found that the weights for states in the chains were low and the
  variances were high, from (3.32607 0.99734) to (731.01504 0.79096).
  On the other hand, for the apnea with a transition to itself, the
  weight and variance was (130965.95661 0.04808).  The classification
  performance was bad.

  I think the sampling rate is too high.  Since the width of the
  low-pass filter for the heart rate data in utilities.py is
  \emph{low\_pass\_width=2 * numpy.pi / (15 * PINT('seconds'))},
  samples that are less than 15/2 seconds apart are redundant.  Forty
  samples per minute is 1.5 seconds per sample or 5 times 15/2.

  I will try changing the sample rate to 8 samples per minute and the
  chain lengths to (4, 8, 1).
\item[Disparate Likelihoods:] The command and last iteration of
  training ar3 are
\begin{verbatim}
python apnea_train.py --AR_order 3 --iterations 20 --records a01 a02 a03 a04 a05 a07 a12 a13 a14 a15 a17 a19 --type masked --trim_start 25 ../../../../build/derived_data/apnea/models/a_ar1_masked ../../../../build/derived_data/apnea/models/a_ar3_masked

19 L0 -6.009e+03 L1  5.868e+03 L2  7.069e+02 L3  4.772e+03 L4  4.285e+03 L5 -1.360e+04 L6  1.951e+03 L7 -2.085e+02 L8 -4.664e+03 L9  3.691e+03 L10 -2.632e+03 L11  7.025e+03 L prior -1.084e+06 U/n -4.9251e+00
\end{verbatim}
  which implies the following likelihoods for individual records:
  
  \begin{tabular*}{1.0\linewidth}{rr} 
    a01 & -6.009e+03 \\
    a02 & 5.868e+03 \\
    a03 & 7.069e+02 \\
    a04 & 4.772e+03 \\
    a05 & 4.285e+03 \\
    a07 & -1.360e+04 \\
    a12 & 1.951e+03 \\
    a13 & -2.085e+02 \\
    a14 & -4.664e+03 \\
    a15 & 3.691e+03 \\
    a17 & -2.632e+03 \\
    a19 & 7.025e+03
  \end{tabular*}
  Also from that training here is a summary of the probability of
  occupying key states averaged over all of the training data:
  
  \begin{tabular*}{1.0\linewidth}[c]{rrllll}
    index & name & weight &  variance & a/b &    alpha \\ \hline
  0 & N\_2\_0     &    2.74e+03  &  0.0449 &  0.2& 5.00e+01 \\
  2 & N\_3\_0     &    1.38e+03  &  0.0847 &  0.2& 5.00e+01 \\
  5 & N\_4\_0     &    1.55e+03  &   0.076 &  0.2& 5.00e+01 \\
  9 & N\_5\_0     &    481       &   0.572 &  0.2& 5.00e+01 \\
 14 & N\_6\_0     &    568       &   0.158 &  0.2& 5.00e+01 \\
 20 & N\_7\_0     &    967       &  0.0665 &  0.2& 5.00e+01 \\
 27 & N\_8\_0     &    3.91e+03  &  0.0272 &  0.2& 5.00e+01 \\
 35 & N\_noise   &    1.31e-18  &     100 &100.0& 9.60e+04 \\
 36 & N\_switch  &    1.17e+04  &  0.0635 &  0.2& 5.00e+01 \\
 37 & A\_2\_0     &    3.92e+03  &  0.0775 &  0.2& 5.00e+01 \\
 39 & A\_3\_0     &    4.11e+03  &  0.0319 &  0.2& 5.00e+01 \\
 42 & A\_4\_0     &    4.4e+03   &  0.0205 &  0.2& 5.00e+01 \\
 46 & A\_5\_0     &    2.71e+03  &  0.0605 &  0.2& 5.00e+01 \\
 51 & A\_6\_0     &    7.86e+03  &  0.0196 &  0.2& 5.00e+01 \\
 57 & A\_7\_0     &    1.9e+03   &  0.0973 &  0.2& 5.00e+01 \\
 64 & A\_8\_0     &    1.28e+03  &   0.585 &  0.2& 5.00e+01 \\
 72 & A\_noise   &    2.28e-17  &     100 &100.0& 9.60e+04 \\
 73 & A\_switch  &    2.63e+04  &   0.059 &  0.2& 5.00e+01
  \end{tabular*}
  I want code to make a similar table for isolated records.

\end{description}

\section{Miscellaneous}
\label{sec:misc}

2023-11-02
\begin{verbatim}
develop-shell$ time make

...

real	37m42.562s
user	319m12.748s
sys	4m8.894s

\end{verbatim}

\end{document}
http://www.ms.k.u-tokyo.ac.jp/sugi/2010/RIMS2010.pdf Density Ratio
Estimation: A Comprehensive Review

https://neuropsychology.github.io/NeuroKit/examples/ecg_delineate/ecg_delineate.html
Locate P, Q, S and T waves in ECG

https://docs.scipy.org/doc/scipy/reference/generated/scipy.datasets.electrocardiogram.html#scipy.datasets.electrocardiogram

http://compbio.fmph.uniba.sk/papers/04wabi.pdf The Most Probable
Labeling Problem in HMMs and Its Application to Bioinformatics

https://probml.github.io/pml-book/book2.html Probabilistic Machine
Learning: Advanced Topics by Kevin Patrick Murphy.  MIT Press, 2023.

https://archive.siam.org/books/textbooks/fr18_book.pdf Iterative
Methods for Optimization C.T. Kelly  (Saved to cathcart)
